lib/pbench/server/api/resources/server_configuration.py

```diff
@@ -113,40 +113,74 @@ def _put(self, params: ApiParams, _) -> Response:
+                keydup.update(params.body.keys())
+            value = params.query.get("value")
+            if value:
+                if params.body:
+                    keydup.update(params.body.keys())
+            else:
+                value = params.body.get("value")
+                if not value:
                     raise APIAbort(
                         HTTPStatus.BAD_REQUEST,
-                        "Incorrect value specification for key {key!r}",
+                        f"No value found for key system configuration key {key!r}",
                     )
+
+            if len(keydup) > 0:
+                raise APIAbort(
+                    HTTPStatus.BAD_REQUEST,
+                    f"Redundant parameters specified in the JSON request body: {sorted(keydup)!r}",
+                )
+
             try:
                 ServerConfig.set(key=key, value=value)
                 return jsonify({key: value})
             except ServerConfigBadValue as e:
                 raise APIAbort(HTTPStatus.BAD_REQUEST, str(e))
             except ServerConfigError as e:
                 raise APIAbort(HTTPStatus.INTERNAL_SERVER_ERROR, str(e))
-        except KeyError:
-            pass
-
-        try:
-            config = params.body["config"]
-        except KeyError:
-            raise MissingParameters(["config"])
 
-        failures = []
-        for k, v in config.items():
-            try:
-                ServerConfig.set(key=k, value=v)
-            except ServerConfigError as e:
-                self.logger.warning("Unable to update key {} = {!r}: {}", k, v, str(e))
-                failures.append(k)
-        if failures:
-            raise APIAbort(HTTPStatus.INTERNAL_SERVER_ERROR)
-        return jsonify(ServerConfig.get_all())
+        else:
+            # If we get here, we didn't get "key" from the URI, so we're expecting
+            # a JSON request body with key/value pairs, potentially for multiple
+            # config options.
+            badkeys = []
+            for k, v in params.body.items():
+                if k not in ServerConfig.KEYS:
+                    badkeys.append(k)
+
+            if badkeys:
+                raise APIAbort(
+                    HTTPStatus.BAD_REQUEST,
+                    f"Unrecognized configuration parameters {sorted(badkeys)!r} specified: valid parameters are {sorted(ServerConfig.KEYS)!r}",
+                )
+
+            failures = []
+            response = {}
+            fail_status = HTTPStatus.BAD_REQUEST
+            for k, v in params.body.items():
+                try:
+                    c = ServerConfig.set(key=k, value=v)
+                    response[c.key] = c.value
+                except ServerConfigBadValue as e:
+                    failures.append(str(e))
+                except ServerConfigError as e:
+                    fail_status = HTTPStatus.INTERNAL_SERVER_ERROR
+                    self.logger.warning(str(e))
+                    failures.append(str(e))
```

Comment:
You don't want to just abort immediately in this case?

Reply:
The problem is we might go through several successful updates before an error (though I assume that's highly unlikely for the general `ServerConfigError` case); so I'd be reporting previous failures but then exiting without trying the remaining keys, and that doesn't seem very consistent.

Reply:
My general thought is that a request should never see a response with an `INTERNAL_SERVER_ERROR`, so, if the Server code is ever inclined to report one, it probably (in general...) doesn't matter what else it might have wanted to do.

Perhaps I misunderstand what is going on in this specific case, but I would think that, if we cannot process one of the keys (due to internal problems), then it really doesn't matter about the other ones.  (Ideally, we would "roll back" any previous updates and just fail the whole request!...but, it's probably not worth setting that up, because the likelihood of this failure arising is really low and the downsides of a partial update are pretty small, so I'm inclined to just take our lumps if this happens...but failing early and immediately makes the code simpler, so I suggest that approach.)